engine/access/subscription/block_tracker.go

```diff
@@ -0,0 +1,128 @@
+// - headers: The storage headers for accessing block headers.
+// - broadcaster: The engine broadcaster for publishing notifications.
+//
+// Returns:
+// - *BlockTrackerImpl: A new instance of BlockTrackerImpl.
+// - error: An error indicating the result of the operation, if any.
+func NewBlockTracker(
+	state protocol.State,
+	rootHeight uint64,
+	headers storage.Headers,
+	broadcaster *engine.Broadcaster,
+) (*BlockTrackerImpl, error) {
+	lastFinalized, err := state.Final().Head()
+	if err != nil {
+		// this header MUST exist in the db, otherwise the node likely has inconsistent state.
+		return nil, irrecoverable.NewExceptionf("could not retrieve last finalized block: %w", err)
+	}
+
+	lastSealed, err := state.Sealed().Head()
+	if err != nil {
+		// this header MUST exist in the db, otherwise the node likely has inconsistent state.
+		return nil, irrecoverable.NewExceptionf("could not retrieve last sealed block: %w", err)
+	}
+
+	return &BlockTrackerImpl{
+		BaseTracker:            NewBaseTrackerImpl(rootHeight, state, headers),
+		state:                  state,
+		finalizedHighestHeight: counters.NewMonotonousCounter(lastFinalized.Height),
+		sealedHighestHeight:    counters.NewMonotonousCounter(lastSealed.Height),
+		broadcaster:            broadcaster,
+	}, nil
+}
+
+// GetHighestHeight returns the highest height based on the specified block status.
+//
+// Parameters:
+// - blockStatus: The status of the block, which could be only BlockStatusSealed or BlockStatusFinalized.
+//
+// No errors are expected during normal operation.
+func (b *BlockTrackerImpl) GetHighestHeight(blockStatus flow.BlockStatus) (uint64, error) {
+	switch blockStatus {
+	case flow.BlockStatusFinalized:
+		return b.finalizedHighestHeight.Value(), nil
+	case flow.BlockStatusSealed:
+		return b.sealedHighestHeight.Value(), nil
+	}
+	return 0, fmt.Errorf("invalid block status: %s", blockStatus)
+}
+
+// ProcessOnFinalizedBlock drives the subscription logic when a block is finalized.
+// The input to this callback is treated as trusted. This method should be executed on
+// `OnFinalizedBlock` notifications from the node-internal consensus instance.
+// No errors are expected during normal operation. Any errors encountered should be
+// treated as an exception.
+func (b *BlockTrackerImpl) ProcessOnFinalizedBlock() error {
+	// get the finalized header from state
+	finalizedHeader, err := b.state.Final().Head()
+	if err != nil {
+		return irrecoverable.NewExceptionf("unable to get latest finalized header: %w", err)
+	}
+
+	if !b.finalizedHighestHeight.Set(finalizedHeader.Height) {
+		return nil
+	}
+
+	// get the latest seal header from state
+	sealedHeader, err := b.state.Sealed().Head()
+	if err != nil {
+		return irrecoverable.NewExceptionf("unable to get latest sealed header: %w", err)
+	}
+
+	// always publish since there is also a new finalized block.
+	_ = b.sealedHighestHeight.Set(sealedHeader.Height)
+
+	b.broadcaster.Publish()
```

Comment:
```suggestion
	_ = b.sealedHighestHeight.Set(sealedHeader.Height)
	// always publish since there is also a new finalized block.
	b.broadcaster.Publish()
```